setup.py

```diff
@@ -49,8 +49,8 @@
 )
 
 setup(
-    name='ptxcompiler',
-    version=versioneer.get_version(),
+    name=f"ptxcompiler{os.getenv('RAPIDS_PY_WHEEL_CUDA_SUFFIX', default='')}",
+    version=os.getenv("RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE", default=versioneer.get_version()),
```

Comment:
This approach works for pypi, but I am worried that the installed package will still use the versioneer-produced version (so there will be a mismatch). Is there a way to test that this doesn't happen?

Reply:
c.f. the discussion here https://github.com/rapidsai/dask-cuda/pull/948 and https://github.com/rapidsai/dask-cuda/issues/336

Reply:
I believe these env vars get "baked" into the wheel at build-time, such that the final `.whl` file reports the correct version

Reply:
What about with `sdist`s?

Reply:
In any event, it would be good to check and confirm that both `whl`s and `sdist`s behave and look correct. Better still would be to have tests of these here.

Reply:
We discussed these issues offline, happy to continue here and get some consensus:
- We have no plans to provide `sdist`s for RAPIDS packages in general, so I think the concern there (while definitely valid) is out of scope for the current effort. 
- In the long run, I am supportive of adding some testing of source dists in order to stress test our build process is robust to various different scenarios. However, I would need a lot of convincing that source distributions are worth actually distributing for RAPIDS packages. We might get away with it in simple cases like `ptxcompiler`, but most of our packages are well above the complexity threshold where sdists will be extremely fragile and depend on details of the local environment. I would prefer to advise users who need that to clone the repositories and build from source, which is effectively what they would be doing anyway but via a far more opaque process (setuptools sdists) that would be extremely challenging to debug and would not provide nearly the level of environmental control as our documented instructions on how to set up a compliant environment environment and configure builds via CMake, etc.
- For nightly versions we are using version overrides that will definitely disagree with what versioneer provides. As a result, `ptxcompiler.__version__` will definitely not be the same as the package metadata version encoded by `setup(...)`. I think that this is an acceptable compromise for this release, but if people feel strongly we could try and resolve it in some manner. The very naive and ugly solution would be to modify the contents of `__init__.py` in a pre-build step to modify the logic for setting `__version__` to not use versioneer. A slightly more sophisticated solution would be to monkey patch `get_versions` in `__init__.py` to respect some static versions.txt file whenever it exists or contains some non-sentinel key, then overwrite that. I don't immediately see a better solution than something like that since we need to capture some build-time generated value and store it into static data that persists when the built wheel is imported.

Reply:
> * For nightly versions we are using version overrides that will definitely disagree with what versioneer provides. As a result, `ptxcompiler.__version__` will definitely not be the same as the package metadata version encoded by `setup(...)`. I think that this is an acceptable compromise for this release, but if people feel strongly we could try and resolve it in some manner. The very naive and ugly solution would be to modify the contents of `__init__.py` in a pre-build step to modify the logic for setting `__version__` to not use versioneer. A slightly more sophisticated solution would be to monkey patch `get_versions` in `__init__.py` to respect some static versions.txt file whenever it exists or contains some non-sentinel key, then overwrite that. I don't immediately see a better solution than something like that since we need to capture some build-time generated value and store it into static data that persists when the built wheel is imported.

I think you don't need to monkey-patch `__init__.py`, but instead provide the override in `setup.py` (e.g. https://github.com/rapidsai/dask-cuda/blob/b3ed9029a1ad02a61eb7fbd899a5a6826bb5cfac/setup.py#L12-L31)

Perhaps I am not understanding something, but this seems to do the right thing for `dask-cuda`'s nightly pip packages (the package metadata and `__version__` match fine).

e.g. for https://files.pythonhosted.org/packages/b6/22/fd7620c923da4b0429adb5d9b7ef782ad089985f9610ff09eb1a6793b225/dask_cuda-23.2.0a221116-py3-none-any.whl

The metadata contains:
```
Metadata-Version: 2.1
Name: dask-cuda
Version: 23.2.0a221116
```

And `_version.py` contains:

```python
import json

version_json = '''
{
 "date": "2022-11-15T14:20:12-0500",
 "dirty": false,
 "error": null,
 "full-revisionid": "3b06fefad640cfa9e00fd8e310b80f2ec6c85d1d",
 "version": "v23.02.00a221116"
}
'''  # END VERSION_JSON


def get_versions():
    return json.loads(version_json)
```

So `__version__` will match.

Reply:
WRT the other two points, those both sound completely fine.

Reply:
This looks promising and is doing effectively what I proposed
> to capture some build-time generated value and store it into static data that persists when the built wheel is imported.

I am not quite sure how this works. Does it completely overwrite the contents of `_version.py`? [The current version on the main branch looks very different](https://github.com/rapidsai/dask-cuda/blob/branch-22.12/dask_cuda/_version.py).

Reply:
That overwriting is handled by the versioneer cmdclass that gets executed `pip wheel` or `pip install` (or whatever) run. So if you're doing `pip install -e .` from a git repo, `_version.py` is left as-is (and `get_versions()` prods git to determine the version). But if you do `pip install .` then `versioneer` rewrites `_version.py` to just contain the baked-in version information.